drivers/adc/adc_shell.c

```diff
@@ -52,178 +82,201 @@ struct adc_hdl {
 		.resolution = 0,					\
 	}
 
-/*
- * TODO generalize with a more flexible for-each macro that doesn't
- * assume a semicolon separator.
- */
-struct adc_hdl adc_list[] = {
-#if DT_NODE_HAS_STATUS(DT_DRV_INST(0), okay)
-	ADC_HDL_LIST_ENTRY(0),
-#endif
-#if DT_NODE_HAS_STATUS(DT_DRV_INST(1), okay)
-	ADC_HDL_LIST_ENTRY(1),
-#endif
-#if DT_NODE_HAS_STATUS(DT_DRV_INST(2), okay)
-	ADC_HDL_LIST_ENTRY(2),
-#endif
+#define INIT_MACRO() DT_INST_FOREACH_STATUS_OKAY(NODE_LABELS) "NA"
+
+/* This table size is = ADC devices count + 1 (NA). */
+static struct adc_hdl {
+	char *device_label;
+	struct adc_channel_cfg channel_config;
+	uint8_t resolution;
+} adc_list[] = {
+	FOR_EACH(ADC_HDL_LIST_ENTRY, (,), INIT_MACRO())
 };
 
-struct args_index {
-	int8_t adc;
-	int8_t parent_adc;
+static const struct args_index {
 	uint8_t channel;
 	uint8_t conf;
 	uint8_t acq_unit;
-};
-
-struct args_number {
-	uint8_t help;
-	uint8_t channel;
-	uint8_t acq_time;
-	uint8_t resolution;
-	uint8_t read;
-};
-
-static const struct args_index args_indx = {
-	.adc = -1,
-	.parent_adc = -2,
+} args_indx = {
 	.channel = 1,
 	.conf = 1,
 	.acq_unit = 2,
 };
 
-static const struct args_number args_no = {
-	.help = 1,
-	.channel = 2,
-	.acq_time = 3,
-	.resolution = 2,
-	.read = 2,
+static const struct gain_string_to_enum {
+	char *string;
+	enum adc_gain gain;
+} gain_list[] = {
+	{ .string = "ADC_GAIN_1_6", .gain = ADC_GAIN_1_6 },
+	{ .string = "ADC_GAIN_1_5", .gain = ADC_GAIN_1_5 },
+	{ .string = "ADC_GAIN_1_4", .gain = ADC_GAIN_1_4 },
+	{ .string = "ADC_GAIN_1_3", .gain = ADC_GAIN_1_3 },
+	{ .string = "ADC_GAIN_1_2", .gain = ADC_GAIN_1_2 },
+	{ .string = "ADC_GAIN_2_3", .gain = ADC_GAIN_2_3 },
+	{ .string = "ADC_GAIN_1", .gain = ADC_GAIN_1 },
+	{ .string = "ADC_GAIN_2", .gain = ADC_GAIN_2 },
+	{ .string = "ADC_GAIN_3", .gain = ADC_GAIN_3 },
+	{ .string = "ADC_GAIN_4", .gain = ADC_GAIN_4 },
+	{ .string = "ADC_GAIN_8", .gain = ADC_GAIN_8 },
+	{ .string = "ADC_GAIN_16", .gain = ADC_GAIN_16 },
+	{ .string = "ADC_GAIN_32", .gain = ADC_GAIN_32 },
+	{ .string = "ADC_GAIN_64", .gain = ADC_GAIN_64 }
```

Comment:
Could you add `ADC_GAIN_128` as well (here and in `sub_gain_cmds`), as it was added in `adc.h`?

Reply:
done